File: 로봇비전/Original Images/dipum_images_ch03/hw2.py
import cv2
import numpy as np
import random
from matplotlib import pyplot as plt

#Read the Images
Im_Pollen = cv2.imread('Fig0308(a)(pollen).tif',cv2.IMREAD_UNCHANGED)
Im_Moon = cv2.imread('Fig0310(a)(Moon Phobos).tif',cv2.IMREAD_UNCHANGED)


#Make Histogram Matrix
Hist_Pollen = cv2.calcHist([Im_Pollen],[0],None,[256],[0,256])
Hist_Moon = cv2.calcHist([Im_Moon],[0],None,[256],[0,256])


#Scale Transformation Function
def Scale(Matrix1):
    Min = np.amin(Matrix1)
    Max = np.amax(Matrix1)
    M1, N1 = Matrix1.shape
    A1 = np.zeros((M1,N1))
    A1 = (255/(Max-Min)*(Matrix1-Min))
    return A1


#Equalize Transform Fuction
def ETF(Histogram1):
    M2, N2 = Histogram1.shape
    Sum = 0
    A2 = np.zeros((M2,N2))
    for i in range(0,M2):
        Sum = Sum + Histogram1[i]
        A2[i] = Sum
    return A2

#Function that helps to make image equilized
def EF(Orig_Image, ETF1):
    M3, N3 = Orig_Image.shape
    A3 = np.zeros((M3,N3))
    for j in range(M3):
        for k in range(N3):
            A3[j, k] = ETF1[Orig_Image[j, k]]
    return A3

#Scale Transformation
SIm_Pollen = np.uint8(Scale(Im_Pollen))
SIm_Moon = np.uint8(Scale(Im_Moon))

#Histogram of Scaled Images
Hist_SPollen = cv2.calcHist([SIm_Pollen],[0],None,[256],[0,256])
Hist_SMoon = cv2.calcHist([SIm_Moon],[0],None,[256],[0,256])


#Make Equalize Transform Function
M4, N4 = Im_Pollen.shape
M5, N5 = Im_Moon.shape
ETF_Pollen = ETF(Hist_Pollen)*255/(M4*N4)
ETF_Moon = ETF(Hist_Moon)*255/(M5*N5)

#Equalize the Images
EF_Pollen = np.uint8(EF(Im_Pollen, ETF_Pollen))
EF_Moon = np.uint8(EF(Im_Moon, ETF_Moon))


#Histogram of Equalized Images
Hist_EFPollen = cv2.calcHist([EF_Pollen],[0],None,[256],[0,256])
Hist_EFMoon = cv2.calcHist([EF_Moon],[0],None,[256],[0,256])

#Show all Images
cv2.imshow('Pollen1', Im_Pollen)
cv2.imshow('Pollen2', SIm_Pollen)
cv2.imshow('Epollen', EF_Pollen)

cv2.imshow('Moon1', Im_Moon)
cv2.imshow('Moon2', SIm_Moon)
cv2.imshow('EMoon', EF_Moon)

#Store all Images
cv2.imwrite('Pollen.tif', Im_Pollen)
cv2.imwrite('SPollen.tif', SIm_Pollen)
cv2.imwrite('EFpollen.tif', EF_Pollen)

cv2.imwrite('Moon.tif', Im_Moon)
cv2.imwrite('SMoon.tif', SIm_Moon)
cv2.imwrite('EFMoon.tif', EF_Moon)


#Show Images and Their Histogram
plt.figure(1)
plt.subplot(221),plt.imshow(Im_Pollen,'gray'),plt.title('Pollen')
plt.subplot(222),plt.imshow(Im_Moon, 'gray'),plt.title('Moon')
plt.subplot(223),plt.plot(Hist_Pollen)
plt.subplot(224),plt.plot(Hist_Moon)
plt.xlim([0,256])
plt.show()

#Compare the Histogram of Orginal Images and Scaled Images
plt.figure(2)
plt.subplot(221),plt.plot(Hist_Pollen),plt.title('Pollen')
plt.subplot(222),plt.plot(Hist_Moon),plt.title('Moon')
plt.subplot(223),plt.plot(Hist_SPollen)
plt.subplot(224),plt.plot(Hist_SMoon)
plt.xlim([0,256])
plt.show()

#compare the Histogram of Original Images and Equalized Images
plt.figure(3)
plt.subplot(221),plt.plot(Hist_Pollen),plt.title('Pollen')
plt.subplot(222),plt.plot(Hist_Moon),plt.title('Moon')
plt.subplot(223),plt.plot(Hist_EFPollen)
plt.subplot(224),plt.plot(Hist_EFMoon)
plt.xlim([0,256])
plt.show()


cv2.waitKey(0)
cv2.destroyAllWindows()

File: 로봇비전/hw2/hw2.m
%%% Robot Vision%%%
%%% Dept. of Electronic Engineering
%%% 201314651 Lee Wonjai

IM_Pollen = imread('C:\Users\user\OneDrive\바탕 화면\2019 특별, 4학년\4학년 2학기\로봇비전\Original Images\dipum_images_ch03\Fig0308(a)(pollen).tif'); % read the targeted image
IM_Moon = imread('C:\Users\user\OneDrive\바탕 화면\2019 특별, 4학년\4학년 2학기\로봇비전\Original Images\dipum_images_ch03\Fig0310(a)(Moon Phobos).tif');


P_Pollen = imhist(IM_Pollen);
P_Moon = imhist(IM_Moon);

%Find Max and Min value of Pollen Image
Min1 = min(find(P_Pollen))-1; %also can be done using min and max value of IM_Pollen
Max1 = max(find(P_Pollen))-1; %the reason of -1 is that array in matlab start to calculate the index of array as 1 

%Find Max and Min value of Moon Image
Min2 = min(find(P_Moon))-1;
Max2 = max(find(P_Moon))-1;

IMScale_Pollen = Scale(IM_Pollen, Max1, Min1);
IMScale_Moon = Scale(IM_Moon, Max2, Min2);

PScale_Pollen = imhist(IMScale_Pollen);
PScale_Moon = imhist(IMScale_Moon);

[m1, n1] = size(IM_Pollen);
[m2, n2] = size(IM_Moon);

ETF_Pollen = ETF(P_Pollen)./(m1.*n1).*255;
ETF_Moon = ETF(P_Moon)./(m2.*n2).*255;

EF_Pollen = uint8(EF(IM_Pollen, ETF_Pollen));
EF_Moon = uint8(EF(IM_Moon, ETF_Moon));

CEF_Pollen = histeq(IM_Pollen, 256);
CEF_Moon = histeq(IM_Moon, 256);

% Compare the histogram of Original Image and scale transformed Image
figure(1);
subplot(1,2,1);
imhist(IM_Pollen)
title('Original Image of Pollen')
subplot(1,2,2);
imhist(IMScale_Pollen)
title('Scale transformed Image of Pollen')


figure(2);
subplot(1,2,1);
imhist(IM_Moon)
title('Original Image of the Moon')
subplot(1,2,2);
imhist(IMScale_Moon)
title('Scale transformed Image of the Moon')

% Compare the Original Image and scale transformed Image
figure(3);
subplot(1,2,1);
imshow(IM_Pollen)
title('Original Image of Pollen')
subplot(1,2,2);
imshow(IMScale_Pollen)
title('Scale transformed Image of Pollen')

figure(4)
subplot(1,2,1);
imshow(IM_Moon)
title('Original Image of the Moon')
subplot(1,2,2);
imshow(IMScale_Moon)
title('Scale transformed Image of the Moon')

% Compare the histogram of Original Image and equalized Image
figure(5);
subplot(1,2,1);
imhist(IM_Pollen)
title('Original Image of Pollen')
subplot(1,2,2);
imhist(EF_Pollen)
title('Equalized Image of Pollen')

figure(6);
subplot(1,2,1);
imhist(IM_Moon)
title('Original Image of the Moon')
subplot(1,2,2);
imhist(EF_Moon)
title('Equalized Image of the Moon')


% Compare the Original Image and equalized Image
figure(7);
subplot(1,2,1);
imshow(IM_Pollen)
title('Original Image of Pollen')
subplot(1,2,2);
imshow(EF_Pollen)
title('Equalized Image of Pollen')

figure(8);
subplot(1,2,1);
imshow(IM_Moon)
title('Original Image of the Moon')
subplot(1,2,2);
imshow(EF_Moon)
title('Equalized Image of the Moon')

%Compare whether equalization that I made is correct or not
figure(9)
subplot(1,2,1);
imhist(EF_Pollen)
title('Equalization Fuction I made')
subplot(1,2,2);
imhist(CEF_Pollen)
title('Equalization Fuction')

figure(10)
subplot(1,2,1);
imhist(EF_Moon)
title('Equalization Fuction I made')
subplot(1,2,2);
imhist(CEF_Moon)
title('Equalization Fuction')

figure(11)
subplot(1,2,1);
imshow(EF_Pollen)
title('Equalization Fuction I made')
subplot(1,2,2);
imshow(CEF_Pollen)
title('Equalization Fuction')

figure(12)
subplot(1,2,1);
imshow(EF_Moon)
title('Equalization Fuction I made')
subplot(1,2,2);
imshow(CEF_Moon)
title('Equalization Fuction')
